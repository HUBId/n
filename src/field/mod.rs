//! Field arithmetic primitives for the `rpp-stark` proof system.
//! Contains finite field implementations and polynomial utilities.

pub mod polynomial;
pub mod prime_field;

pub use prime_field::FieldElement;

#[cfg(test)]
pub mod tests;
