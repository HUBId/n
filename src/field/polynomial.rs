//! Polynomial utilities operating over the prime field.
//! The module provides deterministic arithmetic for trace and constraint polynomials.

use super::FieldElement;

/// Dense polynomial represented by coefficients in ascending order.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Polynomial {
    /// Coefficients starting from the constant term.
    pub coefficients: Vec<FieldElement>,
}

impl Polynomial {
    /// Constructs a polynomial from raw coefficients.
    pub fn new(coefficients: Vec<FieldElement>) -> Self {
        Self { coefficients }
    }

    /// Evaluates the polynomial at the provided point using Horner's method.
    pub fn evaluate(&self, point: FieldElement) -> FieldElement {
        let mut result = FieldElement::zero();
        for coeff in self.coefficients.iter().rev() {
            result = result.mul(point).add(*coeff);
        }
        result
    }

    /// Returns the degree of the polynomial or `None` if the polynomial is zero.
    pub fn degree(&self) -> Option<usize> {
        for (idx, coeff) in self.coefficients.iter().enumerate().rev() {
            if coeff.as_u64() != 0 {
                return Some(idx);
            }
        }
        None
    }
}
