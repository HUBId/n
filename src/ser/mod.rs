//! Canonical serialization helpers for the proof system.
//!
//! The helpers in this module implement the little-endian layouts documented
//! in the repository README. They provide a shared vocabulary for encoding and
//! decoding primitive values and higher-level containers such as `Vec` and
//! `Option`.

mod bytes;
mod collections;
mod cursor;
mod digest;
mod error;
mod felt;
mod ints;

pub use bytes::{
    ensure_consumed, read_exact_bytes, read_prefixed_bytes, write_bytes, write_prefixed_bytes,
};
pub use collections::{read_option, read_vec, write_option, write_vec};
pub use cursor::ByteReader;
pub use digest::{read_digest, write_digest, DIGEST_SIZE};
pub use error::{SerError, SerKind, SerResult};
pub use felt::{read_felt, read_felt_vec, write_felt, write_felt_vec};
pub use ints::{
    ensure_u32, read_bool, read_u128, read_u16, read_u32, read_u64, read_u8, write_bool,
    write_u128, write_u16, write_u32, write_u64, write_u8,
};
