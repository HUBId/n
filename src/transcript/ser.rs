use super::transcript::TranscriptStateView;
use super::types::{SerKind, TranscriptError};

#[derive(Debug, Clone, Copy)]
#[allow(dead_code)]
enum StageTag {
    ExpectPublic = 0,
    TraceRoot = 1,
    TraceChallenge = 2,
    CompRoot = 3,
    CompChallenge = 4,
    FriRoot = 5,
    FriChallenge = 6,
    QueriesNoCount = 7,
    QueriesCount = 8,
    Finalised = 9,
}

#[allow(dead_code)]
impl StageTag {
    fn from_parts(code: u8, aux: u8) -> Option<(StageTag, u8)> {
        let tag = match code {
            0 => StageTag::ExpectPublic,
            1 => StageTag::TraceRoot,
            2 => StageTag::TraceChallenge,
            3 => StageTag::CompRoot,
            4 => StageTag::CompChallenge,
            5 => StageTag::FriRoot,
            6 => StageTag::FriChallenge,
            7 => StageTag::QueriesNoCount,
            8 => StageTag::QueriesCount,
            9 => StageTag::Finalised,
            _ => return None,
        };
        Some((tag, aux))
    }
}

#[allow(dead_code)]
pub(crate) fn serialize_state(view: &TranscriptStateView) -> Vec<u8> {
    let mut out = Vec::with_capacity(48);
    out.extend_from_slice(&view.state);
    out.extend_from_slice(&view.challenge_counter.to_le_bytes());
    match &view.stage {
        super::transcript::Stage::ExpectPublic => {
            out.push(StageTag::ExpectPublic as u8);
            out.push(0);
        }
        super::transcript::Stage::TraceRoot => {
            out.push(StageTag::TraceRoot as u8);
            out.push(0);
        }
        super::transcript::Stage::TraceChallenge => {
            out.push(StageTag::TraceChallenge as u8);
            out.push(0);
        }
        super::transcript::Stage::CompRoot => {
            out.push(StageTag::CompRoot as u8);
            out.push(0);
        }
        super::transcript::Stage::CompChallenge => {
            out.push(StageTag::CompChallenge as u8);
            out.push(0);
        }
        super::transcript::Stage::Fri { layer, expect } => {
            let tag = match expect {
                super::transcript::FriExpectation::Root => StageTag::FriRoot,
                super::transcript::FriExpectation::Challenge => StageTag::FriChallenge,
            };
            out.push(tag as u8);
            out.push(*layer);
        }
        super::transcript::Stage::Queries { count_absorbed } => {
            if *count_absorbed {
                out.push(StageTag::QueriesCount as u8);
            } else {
                out.push(StageTag::QueriesNoCount as u8);
            }
            out.push(0);
        }
        super::transcript::Stage::Finalised => {
            out.push(StageTag::Finalised as u8);
            out.push(0);
        }
    }
    let (phase_code, phase_aux) = match view.phase {
        super::types::TranscriptPhase::Init => (0, 0),
        super::types::TranscriptPhase::Public => (1, 0),
        super::types::TranscriptPhase::TraceCommit => (2, 0),
        super::types::TranscriptPhase::CompCommit => (3, 0),
        super::types::TranscriptPhase::FriLayer(layer) => (4, layer),
        super::types::TranscriptPhase::Queries => (5, 0),
        super::types::TranscriptPhase::Final => (6, 0),
    };
    out.push(phase_code);
    out.push(phase_aux);
    out
}

#[allow(dead_code)]
pub(crate) fn deserialize_state(bytes: &[u8]) -> Result<TranscriptStateView, TranscriptError> {
    if bytes.len() < 42 {
        return Err(TranscriptError::Serialization(SerKind::State));
    }
    let mut state = [0u8; 32];
    state.copy_from_slice(&bytes[..32]);
    let mut counter_bytes = [0u8; 8];
    counter_bytes.copy_from_slice(&bytes[32..40]);
    let challenge_counter = u64::from_le_bytes(counter_bytes);
    let stage_code = *bytes
        .get(40)
        .ok_or(TranscriptError::Serialization(SerKind::State))?;
    let stage_aux = *bytes
        .get(41)
        .ok_or(TranscriptError::Serialization(SerKind::State))?;
    let phase_code = *bytes
        .get(42)
        .ok_or(TranscriptError::Serialization(SerKind::State))?;
    let phase_aux = *bytes.get(43).unwrap_or(&0);

    use super::transcript::{FriExpectation, Stage};
    let stage = match StageTag::from_parts(stage_code, stage_aux) {
        Some((StageTag::ExpectPublic, _)) => Stage::ExpectPublic,
        Some((StageTag::TraceRoot, _)) => Stage::TraceRoot,
        Some((StageTag::TraceChallenge, _)) => Stage::TraceChallenge,
        Some((StageTag::CompRoot, _)) => Stage::CompRoot,
        Some((StageTag::CompChallenge, _)) => Stage::CompChallenge,
        Some((StageTag::FriRoot, layer)) => Stage::Fri {
            layer,
            expect: FriExpectation::Root,
        },
        Some((StageTag::FriChallenge, layer)) => Stage::Fri {
            layer,
            expect: FriExpectation::Challenge,
        },
        Some((StageTag::QueriesNoCount, _)) => Stage::Queries {
            count_absorbed: false,
        },
        Some((StageTag::QueriesCount, _)) => Stage::Queries {
            count_absorbed: true,
        },
        Some((StageTag::Finalised, _)) => Stage::Finalised,
        None => return Err(TranscriptError::Serialization(SerKind::State)),
    };

    use super::types::TranscriptPhase;
    let phase = match (phase_code, phase_aux) {
        (0, _) => TranscriptPhase::Init,
        (1, _) => TranscriptPhase::Public,
        (2, _) => TranscriptPhase::TraceCommit,
        (3, _) => TranscriptPhase::CompCommit,
        (4, layer) => TranscriptPhase::FriLayer(layer),
        (5, _) => TranscriptPhase::Queries,
        (6, _) => TranscriptPhase::Final,
        _ => return Err(TranscriptError::Serialization(SerKind::State)),
    };

    Ok(TranscriptStateView {
        state,
        challenge_counter,
        stage,
        phase,
    })
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::transcript::transcript::{FriExpectation, Stage, TranscriptStateView};
    use crate::transcript::types::TranscriptPhase;

    #[test]
    fn roundtrip_state_encoding() {
        let view = TranscriptStateView {
            state: [42u8; 32],
            challenge_counter: 7,
            stage: Stage::Fri {
                layer: 3,
                expect: FriExpectation::Challenge,
            },
            phase: TranscriptPhase::FriLayer(3),
        };
        let bytes = serialize_state(&view);
        let restored = deserialize_state(&bytes).expect("state decode");
        assert_eq!(restored.state, view.state);
        assert_eq!(restored.challenge_counter, view.challenge_counter);
        match restored.stage {
            Stage::Fri { layer, expect } => {
                assert_eq!(layer, 3);
                assert!(matches!(expect, FriExpectation::Challenge));
            }
            _ => panic!("unexpected stage"),
        }
    }
}
