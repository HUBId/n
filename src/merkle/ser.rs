use super::proof::MerkleProof;
use super::tree::CommitAux;
use super::types::{Digest, MerkleError, ProofNode, SerKind};
use crate::params::{Endianness, MerkleArity};

fn encode_endianness(value: Endianness) -> u8 {
    match value {
        Endianness::Little => 1,
        Endianness::Big => 2,
    }
}

fn decode_endianness(code: u8) -> Option<Endianness> {
    match code {
        1 => Some(Endianness::Little),
        2 => Some(Endianness::Big),
        _ => None,
    }
}

fn encode_arity(arity: MerkleArity) -> u8 {
    match arity {
        MerkleArity::Binary => 2,
        MerkleArity::Quaternary => 4,
    }
}

fn decode_arity(code: u8) -> Option<MerkleArity> {
    match code {
        2 => Some(MerkleArity::Binary),
        4 => Some(MerkleArity::Quaternary),
        _ => None,
    }
}

/// Serialises a [`MerkleProof`] into the canonical byte layout.
pub fn encode_proof(proof: &MerkleProof) -> Result<Vec<u8>, MerkleError> {
    let mut out = Vec::new();
    out.extend_from_slice(&proof.version.to_le_bytes());
    out.push(encode_arity(proof.arity));
    out.push(encode_endianness(proof.leaf_encoding));
    out.push(proof.leaf_width);
    out.extend_from_slice(&proof.domain_sep.to_le_bytes());
    out.extend_from_slice(&proof.leaf_width_bytes.to_le_bytes());
    out.extend_from_slice(&proof.digest_size.to_le_bytes());
    out.extend_from_slice(&(proof.indices.len() as u32).to_le_bytes());
    for &index in &proof.indices {
        out.extend_from_slice(&index.to_le_bytes());
    }
    out.extend_from_slice(&(proof.path.len() as u32).to_le_bytes());
    for node in &proof.path {
        match node {
            ProofNode::Arity2(digests) => {
                out.push(2);
                out.push(digests.len() as u8);
                for digest in digests {
                    if digest.as_bytes().len() != proof.digest_size as usize {
                        return Err(MerkleError::Serialization(SerKind::Proof));
                    }
                    out.extend_from_slice(digest.as_bytes());
                }
            }
            ProofNode::Arity4(digests) => {
                out.push(4);
                out.push(digests.len() as u8);
                for digest in digests {
                    if digest.as_bytes().len() != proof.digest_size as usize {
                        return Err(MerkleError::Serialization(SerKind::Proof));
                    }
                    out.extend_from_slice(digest.as_bytes());
                }
            }
        }
    }
    Ok(out)
}

/// Deserialises a [`MerkleProof`] from its canonical byte representation.
pub fn decode_proof(bytes: &[u8]) -> Result<MerkleProof, MerkleError> {
    let mut cursor = 0usize;
    let mut take = |len: usize| -> Result<&[u8], MerkleError> {
        if cursor + len > bytes.len() {
            return Err(MerkleError::Serialization(SerKind::Proof));
        }
        let slice = &bytes[cursor..cursor + len];
        cursor += len;
        Ok(slice)
    };

    let mut version_bytes = [0u8; 2];
    version_bytes.copy_from_slice(take(2)?);
    let version = u16::from_le_bytes(version_bytes);
    let arity = decode_arity(take(1)?[0]).ok_or(MerkleError::Serialization(SerKind::Proof))?;
    let leaf_encoding =
        decode_endianness(take(1)?[0]).ok_or(MerkleError::Serialization(SerKind::Proof))?;
    let leaf_width = take(1)?[0];
    let mut domain_sep_bytes = [0u8; 8];
    domain_sep_bytes.copy_from_slice(take(8)?);
    let domain_sep = u64::from_le_bytes(domain_sep_bytes);
    let mut leaf_width_bytes_raw = [0u8; 4];
    leaf_width_bytes_raw.copy_from_slice(take(4)?);
    let leaf_width_bytes = u32::from_le_bytes(leaf_width_bytes_raw);
    let mut digest_size_bytes = [0u8; 2];
    digest_size_bytes.copy_from_slice(take(2)?);
    let digest_size = u16::from_le_bytes(digest_size_bytes);
    let mut index_len_bytes = [0u8; 4];
    index_len_bytes.copy_from_slice(take(4)?);
    let index_len = u32::from_le_bytes(index_len_bytes) as usize;
    let mut indices = Vec::with_capacity(index_len);
    for _ in 0..index_len {
        let mut index_bytes = [0u8; 4];
        index_bytes.copy_from_slice(take(4)?);
        indices.push(u32::from_le_bytes(index_bytes));
    }
    let mut path_len_bytes = [0u8; 4];
    path_len_bytes.copy_from_slice(take(4)?);
    let path_len = u32::from_le_bytes(path_len_bytes) as usize;
    let mut path = Vec::with_capacity(path_len);
    for _ in 0..path_len {
        let tag = take(1)?[0];
        let count = take(1)?[0] as usize;
        let expected = match tag {
            2 => 1,
            4 => 3,
            _ => return Err(MerkleError::Serialization(SerKind::Proof)),
        };
        if count != expected {
            return Err(MerkleError::Serialization(SerKind::Proof));
        }
        let mut digests = Vec::with_capacity(count);
        for _ in 0..count {
            let raw = take(digest_size as usize)?;
            digests.push(Digest::new(raw.to_vec()));
        }
        let node = match tag {
            2 => {
                let arr = [digests
                    .into_iter()
                    .next()
                    .ok_or(MerkleError::Serialization(SerKind::Proof))?];
                ProofNode::Arity2(arr)
            }
            4 => {
                let mut iter = digests.into_iter();
                let arr = [
                    iter.next()
                        .ok_or(MerkleError::Serialization(SerKind::Proof))?,
                    iter.next()
                        .ok_or(MerkleError::Serialization(SerKind::Proof))?,
                    iter.next()
                        .ok_or(MerkleError::Serialization(SerKind::Proof))?,
                ];
                ProofNode::Arity4(arr)
            }
            _ => unreachable!(),
        };
        path.push(node);
    }

    Ok(MerkleProof {
        version,
        arity,
        leaf_encoding,
        path,
        indices,
        leaf_width,
        domain_sep,
        leaf_width_bytes,
        digest_size,
    })
}

/// Serialises the commitment auxiliary structure for storage.
pub fn encode_commit_aux(aux: &CommitAux) -> Result<Vec<u8>, MerkleError> {
    let mut out = Vec::new();
    out.extend_from_slice(&aux.version.to_le_bytes());
    out.push(encode_arity(aux.arity));
    out.push(encode_endianness(aux.leaf_encoding));
    out.push(aux.leaf_width);
    out.extend_from_slice(&aux.domain_sep.to_le_bytes());
    out.extend_from_slice(&aux.digest_size.to_le_bytes());
    out.extend_from_slice(&aux.leaf_count.to_le_bytes());
    out.extend_from_slice(&aux.depth.0.to_le_bytes());
    out.extend_from_slice(&(aux.levels.len() as u32).to_le_bytes());
    for level in &aux.levels {
        out.extend_from_slice(&(level.len() as u32).to_le_bytes());
        for digest in level {
            out.extend_from_slice(&(digest.as_bytes().len() as u32).to_le_bytes());
            out.extend_from_slice(digest.as_bytes());
        }
    }
    Ok(out)
}

/// Deserialises [`CommitAux`] from its canonical encoding.
pub fn decode_commit_aux(bytes: &[u8]) -> Result<CommitAux, MerkleError> {
    let mut cursor = 0usize;
    let mut take = |len: usize| -> Result<&[u8], MerkleError> {
        if cursor + len > bytes.len() {
            return Err(MerkleError::Serialization(SerKind::CommitAux));
        }
        let slice = &bytes[cursor..cursor + len];
        cursor += len;
        Ok(slice)
    };

    let mut version_bytes = [0u8; 2];
    version_bytes.copy_from_slice(take(2)?);
    let version = u16::from_le_bytes(version_bytes);
    let arity = decode_arity(take(1)?[0]).ok_or(MerkleError::Serialization(SerKind::CommitAux))?;
    let leaf_encoding =
        decode_endianness(take(1)?[0]).ok_or(MerkleError::Serialization(SerKind::CommitAux))?;
    let leaf_width = take(1)?[0];
    let mut domain_sep_bytes = [0u8; 8];
    domain_sep_bytes.copy_from_slice(take(8)?);
    let domain_sep = u64::from_le_bytes(domain_sep_bytes);
    let mut digest_size_bytes = [0u8; 2];
    digest_size_bytes.copy_from_slice(take(2)?);
    let digest_size = u16::from_le_bytes(digest_size_bytes);
    let mut leaf_count_bytes = [0u8; 4];
    leaf_count_bytes.copy_from_slice(take(4)?);
    let leaf_count = u32::from_le_bytes(leaf_count_bytes);
    let mut depth_bytes = [0u8; 4];
    depth_bytes.copy_from_slice(take(4)?);
    let depth = u32::from_le_bytes(depth_bytes);
    let mut level_count_bytes = [0u8; 4];
    level_count_bytes.copy_from_slice(take(4)?);
    let level_count = u32::from_le_bytes(level_count_bytes) as usize;
    let mut levels = Vec::with_capacity(level_count);
    for _ in 0..level_count {
        let mut node_count_bytes = [0u8; 4];
        node_count_bytes.copy_from_slice(take(4)?);
        let node_count = u32::from_le_bytes(node_count_bytes) as usize;
        let mut level = Vec::with_capacity(node_count);
        for _ in 0..node_count {
            let mut len_bytes = [0u8; 4];
            len_bytes.copy_from_slice(take(4)?);
            let len = u32::from_le_bytes(len_bytes) as usize;
            let raw = take(len)?;
            level.push(Digest::new(raw.to_vec()));
        }
        levels.push(level);
    }

    Ok(CommitAux {
        version,
        arity,
        leaf_width,
        leaf_encoding,
        domain_sep,
        digest_size,
        leaf_count,
        depth: super::types::TreeDepth(depth),
        levels,
    })
}
