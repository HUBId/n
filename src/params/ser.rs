use crate::ser::{
    ensure_consumed, read_exact_bytes, read_u16, read_u32, read_u64, read_u8, write_bytes,
    write_u16, write_u32, write_u64, write_u8, ByteReader, SerError, SerKind,
};

use super::types::{
    ChallengeBounds, Endianness, FieldKind, FriFolding, FriParams, HashFamily, HashKind, LdeOrder,
    LdeParams, MerkleArity, MerkleParams, ProofParams, SecurityBudget, TranscriptParams,
};
use super::StarkParams;

const PARAMS_SER_KIND: SerKind = SerKind::Params;

/// Canonical binary serialisation for [`StarkParams`].
pub fn serialize_params(params: &StarkParams) -> Vec<u8> {
    let mut out = Vec::with_capacity(89);
    write_u16(&mut out, params.params_version());
    write_u16(&mut out, params.field().code());
    write_u8(&mut out, params.hash().family().code());
    write_u16(&mut out, params.hash().parameter_id());
    write_u32(&mut out, params.lde().blowup);
    write_u8(&mut out, params.lde().order.code());
    write_u64(&mut out, params.lde().coset_tag);
    write_u8(&mut out, params.fri().r);
    write_u16(&mut out, params.fri().queries);
    write_u16(&mut out, params.fri().domain_log2);
    write_u8(&mut out, params.fri().folding.code());
    write_u8(&mut out, params.fri().num_layers);
    write_u8(&mut out, params.merkle().leaf_encoding.code());
    write_u8(&mut out, params.merkle().leaf_width);
    write_u8(&mut out, params.merkle().arity.code());
    write_u64(&mut out, params.merkle().domain_sep);
    write_u64(&mut out, params.transcript().protocol_tag);
    write_bytes(&mut out, &params.transcript().seed);
    write_u8(&mut out, params.transcript().challenge_bounds.minimum);
    write_u8(&mut out, params.transcript().challenge_bounds.maximum);
    write_u16(&mut out, params.proof().version);
    write_u32(&mut out, params.proof().max_size_kb);
    write_u16(&mut out, params.security().target_bits);
    write_u8(&mut out, params.security().soundness_slack_bits);
    out
}

/// Deserialises a parameter set from canonical bytes.
pub fn deserialize_params(bytes: &[u8]) -> Result<StarkParams, SerError> {
    let mut cursor = ByteReader::new(bytes);

    let params_version = read_u16(&mut cursor, PARAMS_SER_KIND, "params_version")?;
    let field_code = read_u16(&mut cursor, PARAMS_SER_KIND, "field")?;
    let field = FieldKind::from_code(field_code)
        .ok_or_else(|| SerError::invalid_value(PARAMS_SER_KIND, "field"))?;

    let family_code = read_u8(&mut cursor, PARAMS_SER_KIND, "hash.family")?;
    let family = HashFamily::from_code(family_code)
        .ok_or_else(|| SerError::invalid_value(PARAMS_SER_KIND, "hash.family"))?;
    let hash_param = read_u16(&mut cursor, PARAMS_SER_KIND, "hash.parameter_id")?;
    let hash = HashKind::from_codes(family, hash_param);

    let lde_blowup = read_u32(&mut cursor, PARAMS_SER_KIND, "lde.blowup")?;
    let lde_order_code = read_u8(&mut cursor, PARAMS_SER_KIND, "lde.order")?;
    let lde_order = LdeOrder::from_code(lde_order_code)
        .ok_or_else(|| SerError::invalid_value(PARAMS_SER_KIND, "lde.order"))?;
    let lde_coset = read_u64(&mut cursor, PARAMS_SER_KIND, "lde.coset_tag")?;

    let fri_r = read_u8(&mut cursor, PARAMS_SER_KIND, "fri.r")?;
    let fri_queries = read_u16(&mut cursor, PARAMS_SER_KIND, "fri.queries")?;
    let fri_domain = read_u16(&mut cursor, PARAMS_SER_KIND, "fri.domain_log2")?;
    let fri_folding_code = read_u8(&mut cursor, PARAMS_SER_KIND, "fri.folding")?;
    let fri_folding = FriFolding::from_code(fri_folding_code)
        .ok_or_else(|| SerError::invalid_value(PARAMS_SER_KIND, "fri.folding"))?;
    let fri_layers = read_u8(&mut cursor, PARAMS_SER_KIND, "fri.num_layers")?;

    let merkle_encoding_code = read_u8(&mut cursor, PARAMS_SER_KIND, "merkle.leaf_encoding")?;
    let merkle_encoding = Endianness::from_code(merkle_encoding_code)
        .ok_or_else(|| SerError::invalid_value(PARAMS_SER_KIND, "merkle.leaf_encoding"))?;
    let merkle_leaf_width = read_u8(&mut cursor, PARAMS_SER_KIND, "merkle.leaf_width")?;
    let merkle_arity_code = read_u8(&mut cursor, PARAMS_SER_KIND, "merkle.arity")?;
    let merkle_arity = MerkleArity::from_code(merkle_arity_code)
        .ok_or_else(|| SerError::invalid_value(PARAMS_SER_KIND, "merkle.arity"))?;
    let merkle_domain_sep = read_u64(&mut cursor, PARAMS_SER_KIND, "merkle.domain_sep")?;

    let transcript_protocol = read_u64(&mut cursor, PARAMS_SER_KIND, "transcript.protocol_tag")?;
    let transcript_seed = {
        let bytes = read_exact_bytes(&mut cursor, PARAMS_SER_KIND, "transcript.seed", 32)?;
        let mut seed = [0u8; 32];
        seed.copy_from_slice(bytes);
        seed
    };
    let transcript_ch_min = read_u8(
        &mut cursor,
        PARAMS_SER_KIND,
        "transcript.challenge_bounds.minimum",
    )?;
    let transcript_ch_max = read_u8(
        &mut cursor,
        PARAMS_SER_KIND,
        "transcript.challenge_bounds.maximum",
    )?;

    let proof_version = read_u16(&mut cursor, PARAMS_SER_KIND, "proof.version")?;
    let proof_max_size = read_u32(&mut cursor, PARAMS_SER_KIND, "proof.max_size_kb")?;
    let security_target = read_u16(&mut cursor, PARAMS_SER_KIND, "security.target_bits")?;
    let security_slack = read_u8(
        &mut cursor,
        PARAMS_SER_KIND,
        "security.soundness_slack_bits",
    )?;

    ensure_consumed(&cursor, PARAMS_SER_KIND)?;

    Ok(StarkParams {
        params_version,
        field,
        hash,
        lde: LdeParams {
            blowup: lde_blowup,
            order: lde_order,
            coset_tag: lde_coset,
        },
        fri: FriParams {
            r: fri_r,
            queries: fri_queries,
            domain_log2: fri_domain,
            folding: fri_folding,
            num_layers: fri_layers,
        },
        merkle: MerkleParams {
            leaf_encoding: merkle_encoding,
            leaf_width: merkle_leaf_width,
            arity: merkle_arity,
            domain_sep: merkle_domain_sep,
        },
        transcript: TranscriptParams {
            protocol_tag: transcript_protocol,
            seed: transcript_seed,
            challenge_bounds: ChallengeBounds {
                minimum: transcript_ch_min,
                maximum: transcript_ch_max,
            },
        },
        proof: ProofParams {
            version: proof_version,
            max_size_kb: proof_max_size,
        },
        security: SecurityBudget {
            target_bits: security_target,
            soundness_slack_bits: security_slack,
        },
    })
}
