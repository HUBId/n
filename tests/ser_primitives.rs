use insta::assert_snapshot;
use rpp_stark::field::FieldElement;
use rpp_stark::ser::{
    read_bool, read_digest, read_felt, read_prefixed_bytes, read_u16, read_u32, write_bool,
    write_digest, write_felt, write_prefixed_bytes, write_u16, write_u32, write_u8, ByteReader,
    SerError, SerKind,
};

#[test]
fn roundtrip_unsigned_integers() {
    let mut buffer = Vec::new();
    write_u16(&mut buffer, 0x1234);
    write_u32(&mut buffer, 0xdead_beef);
    let mut cursor = ByteReader::new(&buffer);
    assert_eq!(
        read_u16(&mut cursor, SerKind::Proof, "u16").unwrap(),
        0x1234
    );
    assert_eq!(
        read_u32(&mut cursor, SerKind::Proof, "u32").unwrap(),
        0xdead_beef
    );
    assert_eq!(cursor.remaining(), 0);
}

#[test]
fn bool_roundtrip_and_invalid() {
    let mut buffer = Vec::new();
    write_bool(&mut buffer, true);
    write_bool(&mut buffer, false);
    let mut cursor = ByteReader::new(&buffer);
    assert!(read_bool(&mut cursor, SerKind::Proof, "flag").unwrap());
    assert!(!read_bool(&mut cursor, SerKind::Proof, "flag").unwrap());

    let invalid = [2u8];
    let mut cursor = ByteReader::new(&invalid);
    let err = read_bool(&mut cursor, SerKind::Proof, "flag").expect_err("invalid flag");
    assert!(matches!(err, SerError::InvalidValue { .. }));
}

#[test]
fn felt_roundtrip() {
    let element = FieldElement(0xfeed_beefu64);
    let mut buffer = Vec::new();
    write_felt(&mut buffer, element);
    let mut cursor = ByteReader::new(&buffer);
    let restored = read_felt(&mut cursor, SerKind::Proof, "felt").unwrap();
    assert_eq!(restored, element);
}

#[test]
fn length_prefixed_bytes_roundtrip() {
    let payload = [1u8, 2, 3, 4, 5];
    let mut buffer = Vec::new();
    write_prefixed_bytes(&mut buffer, &payload, SerKind::PublicInputs, "bytes").unwrap();
    let mut cursor = ByteReader::new(&buffer);
    let restored = read_prefixed_bytes(&mut cursor, SerKind::PublicInputs, "bytes").unwrap();
    assert_eq!(restored, payload);
}

#[test]
fn snapshot_digest_encoding() {
    let mut buffer = Vec::new();
    write_digest(&mut buffer, &[0xabu8; 32]);
    write_u8(&mut buffer, 7);
    let hex = buffer
        .iter()
        .map(|byte| format!("{:02x}", byte))
        .collect::<Vec<_>>()
        .join(" ");
    assert_snapshot!("digest_with_tag", hex);
}

#[test]
fn read_prefixed_bytes_short_buffer() {
    let data = [0x05, 0x00, 0x00, 0x00, 1, 2, 3];
    let mut cursor = ByteReader::new(&data);
    let err =
        read_prefixed_bytes(&mut cursor, SerKind::Telemetry, "payload").expect_err("short buffer");
    assert!(matches!(err, SerError::UnexpectedEnd { .. }));
}

#[test]
fn read_digest_truncated() {
    let mut cursor = ByteReader::new(&[0u8; 16]);
    let err = read_digest(&mut cursor, SerKind::Params, "digest").expect_err("truncated digest");
    assert!(matches!(err, SerError::UnexpectedEnd { .. }));
}
