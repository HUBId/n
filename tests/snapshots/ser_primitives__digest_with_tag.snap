---
source: tests/ser_primitives.rs
assertion_line: 71
expression: hex
---
ab ab ab ab ab ab ab ab ab ab ab ab ab ab ab ab ab ab ab ab ab ab ab ab ab ab ab ab ab ab ab ab 07
